### PR TITLE
Onboard postgresflex backup commands

### PR DESCRIPTION
**Pull request by GokceGK**
Following commands have been added

instance
  - backups
     -  list
     - describe
     - update-schedule

### CODE REVIEW

File: internal/cmd/postgresflex/instance/backups/describe/describe.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+			if err != nil {
+				return err
+			}
+
+			// Call API
+			req := buildRequest(ctx, model, apiClient)
+			resp, err := req.Execute()
+
+			if err != nil {
+				return fmt.Errorf("describe backup for PostgreSQL Flex instance : %w", err)
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/describe/describe.go:66]:
```suggestion
				return fmt.Errorf("describe backup for PostgreSQL Flex instance: %w", err)
```

File: internal/cmd/postgresflex/instance/backups/describe/describe.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	req := apiClient.GetBackup(ctx, model.ProjectId, model.InstanceId, model.BackupId)
+	return req
+}
+
+func outputResult(cmd *cobra.Command, outputFormat string, backup postgresflex.Backup) error {
+	backupStartTime, err := time.Parse(time.RFC3339, *backup.StartTime)
+	if err != nil {
+		return fmt.Errorf("parse backup start time : %w", err)
+	}
+	backupExpireDate := backupStartTime.AddDate(0, 0, 30).Format(time.DateOnly)
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/describe/describe.go:108]: this shouldn't be hardcoded, at least make it a constant

Also, where did you get the value from? Is the backup always lasting 30 days?

File: internal/cmd/postgresflex/instance/backups/describe/describe.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		err := table.Display(cmd)
+		if err != nil {
+			return fmt.Errorf("render table: %w", err)
+		}
+
+		return nil
+	default:
+		details, err := json.MarshalIndent(backup, "", "  ")
+		if err != nil {
+			return fmt.Errorf("marshal backup for PostgreSQL Flex instance : %w", err)
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/describe/describe.go:134]:
```suggestion
			return fmt.Errorf("marshal backup for PostgreSQL Flex instance: %w", err)
```

File: internal/cmd/postgresflex/instance/backups/describe/describe.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		Long:  "Shows details of a backup for a specific PostgreSQL Flex instance.",
+		Example: examples.Build(
+			examples.NewExample(
+				`Get details of a backup with ID "xxx" for a PostgreSQL Flex instance with ID "yyy"`,
+				"$ stackit postgresflex instance backups describe xxx --instance-id yyy"),
+			examples.NewExample(
+				`Get details of a backup with ID "xxx" for a PostgreSQL Flex instance with ID "yyy" in a table format`,
+				"$ stackit postgresflex instance backups describe xxx --instance-id yyy --output-format pretty"),
+		),
+		Args: args.SingleArg(backupIdArg, nil),
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/describe/describe.go:47]:
```suggestion
		Args: args.SingleArg(backupIdArg, utils.ValidateUUID)),
```

We should validate that the provided backup id is a UUID

File: internal/cmd/postgresflex/instance/backups/describe/describe_test.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				flagValues[instanceIdFlag] = "invalid-uuid"
+			}),
+			isValid: false,
+		},
+		{
+			description: "backup id invalid 1",
+			argValues:   []string{""},
+			flagValues:  fixtureFlagValues(),
+			isValid:     false,
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/describe/describe_test.go:153]: add a test case where backup id is a non-empty string with an invalid uuid

File: internal/cmd/postgresflex/instance/backups/describe/describe.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		table.AddSeparator()
+		table.AddRow("NAME", *backup.Name)
+		table.AddSeparator()
+		table.AddRow("START TIME", *backup.StartTime)
+		table.AddSeparator()
+		table.AddRow("END TIME", *backup.EndTime)
+		table.AddSeparator()
+		table.AddRow("EXPIRES AT", backupExpireDate)
+		table.AddSeparator()
+		table.AddRow("BACKUP SIZE", bytesize.Format(uint64(*backup.Size)))
```
[the remark this answers is not in the document]
Comment (vicentepinto98) [reply internal/cmd/postgresflex/instance/backups/describe/describe.go:123]: Ahh there is no Portal integration yet right?

File: internal/cmd/postgresflex/instance/backups/list/list.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		details, err := json.MarshalIndent(backups, "", "  ")
+		if err != nil {
+			return fmt.Errorf("marshal PostgreSQL Flex instance list: %w", err)
+		}
+		cmd.Println(string(details))
+
+		return nil
+	default:
+		table := tables.NewTable()
+		table.SetHeader("ID", "NAME", "START TIME", "END TIME", "BACKUP SIZE")
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/list/list.go:140]: there is no Portal integration yet, but I would say to add also the `EXPIRES_AT`, unless the table output becomes too big

File: internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/stackitcloud/stackit-cli/internal/pkg/flags"
+	"github.com/stackitcloud/stackit-cli/internal/pkg/globalflags"
+	"github.com/stackitcloud/stackit-cli/internal/pkg/services/postgresflex/client"
+	postgresflexUtils "github.com/stackitcloud/stackit-cli/internal/pkg/services/postgresflex/utils"
+	"github.com/stackitcloud/stackit-sdk-go/services/postgresflex"
+)
+
+const (
+	instanceIdFlag     = "instance-id"
+	backupScheduleFlag = "backup-schedule"
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go:21]:
```suggestion
	scheduleFlag = "schedule"
```
`backup` is redundant

File: internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+			return nil
+		},
+	}
+	configureFlags(cmd)
+	return cmd
+}
+
+func configureFlags(cmd *cobra.Command) {
+	cmd.Flags().Var(flags.UUIDFlag(), instanceIdFlag, "Instance ID")
+	cmd.Flags().String(backupScheduleFlag, "", "Backup schedule, in the cron scheduling system format e.g. '0 0 * * *'")
```
Comment (vicentepinto98) [inline internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go:87]: how does the API error look like if the schedule is provided in a wrong format?

File: internal/cmd/postgresflex/instance/instance.go [hunk 10]
```diff
@@ -32,4 +33,5 @@ func addSubcommands(cmd *cobra.Command) {
 	cmd.AddCommand(update.NewCmd())
 	cmd.AddCommand(delete.NewCmd())
 	cmd.AddCommand(clone.NewCmd())
+	cmd.AddCommand(backups.NewCmd())
```
[the remark this answers is not in the document]
Comment (joaopalet) [reply internal/cmd/postgresflex/instance/instance.go:36]: I though it made more sense to have it under the `instance` command group as the backups are specific on an instance. I checked how others did it and it seems to be done both ways.

But now thinking about it, for the DSA services for example we also have the `credentials` command group under the `<service>` group even though the credentials belong to an instance

[the remark this answers is not in the document]
Comment (joaopalet) [reply internal/cmd/postgresflex/instance/instance.go:36]: So yes, for consistency maybe we could move it to be a subcommand of `postgresflex` (sorry @GokceGK for having to move it again!)

[the remark this answers is not in the document]
Comment (vicentepinto98) [reply internal/cmd/postgresflex/instance/instance.go:36]: Exactly, it's the same with `record-set` for example, it also belongs to a `zone` but we decided to have only one command group as the standard

File: internal/cmd/postgresflex/instance/backups/describe/describe.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (GokceGK) [reply internal/cmd/postgresflex/instance/backups/describe/describe.go:123]: No, we dont yet have integration. I checked MongoDB Flex integration and in there it was converted, that is why I convert here, only bytes is not really readable. If needed, we might consider another library. We will also need at least for MongoDB integration.

File: internal/cmd/postgresflex/instance/backups/describe/describe.go (hunk 4 above)
Comment (GokceGK) [reply internal/cmd/postgresflex/instance/backups/describe/describe.go:47]: Backup ID does not have to be a UUID, that is why I did not add validation. 
I also verified it now with the Postgres team.

File: internal/cmd/postgresflex/instance/backups/describe/describe_test.go (hunk 5 above)
Comment (GokceGK) [reply internal/cmd/postgresflex/instance/backups/describe/describe_test.go:153]: I think this will not be needed as I can not validate backup id as UUID. Please see my comment from previous conversation.

File: internal/cmd/postgresflex/instance/backups/list/list.go (hunk 7 above)
Comment (GokceGK) [reply internal/cmd/postgresflex/instance/backups/list/list.go:140]: I will add it. 
I was looking at the backups list output in portal for MongoDB flex, I think I will use the columns, they have used in portal.  
Besides, I plan to use the same format for DateTimes like they use in table.

File: internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go (hunk 9 above)
Comment (GokceGK) [reply internal/cmd/postgresflex/instance/backups/update-schedule/update_schedule.go:87]: could not test this yet for PostgreSQL flex, but MongoDB Flex returns "invalid format"

File: internal/cmd/postgresflex/instance/backups/describe/describe.go (hunk 2 above)
Comment (vicentepinto98) [reply internal/cmd/postgresflex/instance/backups/describe/describe.go:108]: Ok, then just make it a const for better maitainability

File: internal/cmd/postgresflex/instance/backups/describe/describe.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vicentepinto98) [reply internal/cmd/postgresflex/instance/backups/describe/describe.go:123]: Alright, the library itself doesn't seem nothing special, not introducing some vulnerability. I was just intrigued by it not being used at all apparently.

I found this one that at least has some usage: [go-bytesize](https://github.com/inhies/go-bytesize)

File: internal/cmd/postgresflex/backup/describe/describe.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+	InstanceId string
+	BackupId   string
+}
+
+func NewCmd(p *print.Printer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("describe %s", backupIdArg),
+		Short: "Shows details of a backup for a specific PostgreSQL Flex instance",
+		Long:  "Shows details of a backup for a specific PostgreSQL Flex instance.",
```
[the remark this answers is not in the document]
Comment (DiogoFerrao) [reply internal/cmd/postgresflex/backup/describe/describe.go:44]: I think we can drop the "specific" for consistency with other commands

File: internal/cmd/postgresflex/backup/list/list.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+	InstanceId *string
+	Limit      *int64
+}
+
+func NewCmd(p *print.Printer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "Lists all backups which are available for a specific PostgreSQL Flex instance",
+		Long:  "Lists all backups which are available for a specific PostgreSQL Flex instance.",
```
Comment (DiogoFerrao) [inline internal/cmd/postgresflex/backup/list/list.go:44]:
```suggestion
		Short: "Lists all backups which are available for a PostgreSQL Flex instance",
		Long:  "Lists all backups which are available for a PostgreSQL Flex instance.",
```

Comment (DiogoFerrao) [reply internal/cmd/postgresflex/backup/list/list.go:44]: Same here

File: internal/cmd/postgresflex/backup/describe/describe.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+	switch outputFormat {
+	case print.PrettyOutputFormat:
+		table := tables.NewTable()
+		table.AddRow("ID", *backup.Id)
+		table.AddSeparator()
+		table.AddRow("START TIME", *backup.StartTime)
+		table.AddSeparator()
+		table.AddRow("EXPIRES AT", backupExpireDate)
+		table.AddSeparator()
+		table.AddRow("BACKUP SIZE", bytesize.New(float64(*backup.Size)))
```
Comment (DiogoFerrao) [inline internal/cmd/postgresflex/backup/describe/describe.go:125]: Can't we show more info? I'm not sure if this was clarified with the service team, but the response has Labels and Options, which could be useful.
Only problem I see is that describe is returning the same as list, which would make this a bit useless

File: internal/cmd/postgresflex/backup/list/list_test.go [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+			}),
+			isValid: false,
+		},
+		{
+			description: "instance id invalid 1",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				flagValues[instanceIdFlag] = ""
+			}),
+			isValid: false,
+		},
```
Comment (DiogoFerrao) [inline internal/cmd/postgresflex/backup/list/list_test.go:112]: Missing an invalid uuid test for instance id:
```
		{
			description: "instance id invalid 2",
			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
				flagValues[instanceIdFlag] = "invalid-uuid"
			}),
			isValid: false,
		},
```

File: internal/cmd/postgresflex/backup/update-schedule/update_schedule.go [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+	InstanceId     *string
+	BackupSchedule *string
+}
+
+func NewCmd(p *print.Printer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "update-schedule",
+		Short: "Updates backup schedule for a specific PostgreSQL Flex instance",
+		Long:  "Updates backup schedule for a specific PostgreSQL Flex instance.",
```
[the remark this answers is not in the document]
Comment (DiogoFerrao) [reply internal/cmd/postgresflex/backup/update-schedule/update_schedule.go:35]: same here

File: internal/cmd/postgresflex/backup/update-schedule/update_schedule_test.go [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+			}),
+			isValid: false,
+		},
+		{
+			description: "instance id invalid 1",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				flagValues[instanceIdFlag] = ""
+			}),
+			isValid: false,
+		},
```
Comment (DiogoFerrao) [inline internal/cmd/postgresflex/backup/update-schedule/update_schedule_test.go:125]: missing invalid instance 2 (same as list)

File: internal/cmd/postgresflex/backup/describe/describe.go (hunk 13 above)
Comment (GokceGK) [reply internal/cmd/postgresflex/backup/describe/describe.go:125]: Good point, I will ping the Postgres team and ask

**Review by DiogoFerrao — APPROVED** [review]

File: internal/cmd/postgresflex/backup/describe/describe.go (hunk 13 above)
Comment (DiogoFerrao) [reply internal/cmd/postgresflex/backup/describe/describe.go:125]: That's all the information they return, then I think this is a non-issue

**Review by DiogoFerrao — APPROVED** [review]
